### DISCUSSION TITLE
Config settings for gradual adoption

### DISCUSSION THREAD

**Discussion by jrheard-seesaw** [microsoft/pyright#8034] [Q&A] [answered]
Hi! I'm interested in trying out Pyright+Pylance. I'm trying to figure out a plan for gradual adoption in a preexisting codebase; I basically want to set `typeCheckingMode: "standard"` for a small handful of (potentially deeply nested) directories to start off, and then gradually grow that list of directories over time.

In mypy, I would do something like this in my `mypy.ini`:

```ini
[mypy]
strict = True
ignore_errors = True

[mypy-util.some_directory.*]
ignore_errors = False

[mypy-backend.models.validation.*]
ignore_errors = False
```

I'm trying to figure out how to achieve a similar setup with Pyright/Pylance.

I have a `pyrightconfig.json` that looks something like this:

```json
{
    "include": [
        "backend/models/validation"
    ],
    "pythonVersion": "3.11",
    "pythonPlatform": "Linux",
    "typeCheckingMode": "standard",
}
```

That config file causes command-line Pyright to behave the way I want, but when editing files in some other directory (eg `util/some_directory/foo.py`), the Pylance extension is reporting type errors that I don't want to see because I'm not yet ready to fix them.

What should I do differently to cause Pylance and Pyright to both behave the same way (i.e. not report type errors for any directories **except** the ones listed in my `"include"` list)?

Thank you!

**Comment by rchiodo** [comment]
I believe that should work except that when you open a file in the IDE, it's automatically included. Open files count towards what we analyze.

You could also add an ignore section, but then you get no diagnostics for the `util` folder:

```
{
    "include": [
        "backend/models/validation"
    ],
    "ignore": ["util"],
    "pythonVersion": "3.11",
    "pythonPlatform": "Linux",
    "typeCheckingMode": "standard",
}
```

I think what you really want is this enhancement request:

https://github.com/microsoft/pylance-release/issues/5564

That way you could have different pyrightconfig.json files at different folders to set the typecheckingmode for each.

**Reply by jrheard-seesaw** [reply]
Thanks for the suggestions! That ignore section approach is probably not going to work for us, because I'll also want to start getting individual subdirectories like `util/foo/bar` to pass typechecking one at a time for quite some time before we get to the point where the entire `util` directory is ready to be type checked.

As someone new to Pyright, a workspace-based approach seems a bit heavyweight conceptually, because it's not the case that each of these subdirectories is its own logical workspace; it's just the case that we have a ton of directories in a single not-massive-but-not-small Python repo, and we want to gradually introduce type checking into our project one small deeply-nested directory at a time.

(Am I going about this in the wrong way? How do folks usually do gradual adoption of Pyright/Pylance in a preexisting codebase?)

That said, I'll take any solution I can get :) I'd really love for the in-editor Pylance experience to match the command-line Pyright experience. If I'm not able to achieve that parity, I'll probably have to go with Mypy instead of Pyright for our project. 🤞

**Reply by rchiodo** [reply]
The ignore [section](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options) supports globs, so you could add specific sections if that worked for you:

```
   "ignore" : [ "util/**/bar", "util/**/baz"]
```

> How do folks usually do gradual adoption of Pyright/Pylance in a preexisting codebase?

@erictraut will probably have more feedback, but I've seen is people start with typeCheckingMode=off in the IDE and have a different configuration for CI. So they turn typecheckingmode to strict in the IDE temporarily as they work on a specific subdirectory and then in their CI pyrightconfig.json, they forcefully include those directories so they don't regress.

Something like what @erictraut describes here:
https://github.com/microsoft/pyright/issues/6240#issuecomment-1782264981

**Reply by rchiodo** [reply]
It's not ideal though because then everybody else can introduce strict mode bugs in the IDE and they don't realize it until they submit to CI. I think the ideal solution would be for us to have support for pyrightconfig.json in multiple directories.

**Reply by jrheard-seesaw** [reply]
The glob idea is interesting! Do you think that an approach like this could work?


```
    "include": [
        "backend/models/validation",
        "util/foo/bar",
    ],
    "ignore": [
        "backend/!(models/validation)/**",
        "util/!(foo/bar)/**"
    ],
```

I'm not super skilled with glob patterns, but from some poking around so far it seems like some tools would interpret globs like that in a way that would accomplish my goal. It looks like [the relevant pyright code](https://github.com/microsoft/pyright/blob/e69974cee01cd70e50b4c1d0559d052a8176311d/packages/pyright-internal/src/common/uri/uriUtils.ts#L210) (assuming that I found the right code) doesn't implement this `!` behavior, so I guess that won't be an option for me here.

(Please let me know if my glob syntax is incorrect or if I'm totally mistaken about the ability for this to be expressed in any known version of glob syntax!)

Do you think that it could be worthwhile to open an issue asking for pyright to support exclamation points in glob patterns seen in its config file? Or is my use case too niche / am I approaching this from the wrong angle?

**Reply by rchiodo** [reply]
I would guess that Eric would say to use `include` patterns to implement the `!` patterns for ignore. Like he mentions here:
https://github.com/microsoft/pyright/issues/6851 but that doesn't seem to work for ignore. Ignore seems to trump include, which I guess makes sense. 

So I think you'd need ! support for at least ignore.

**Reply by erictraut** [reply, marked as answer]
For general guidance in applying gradual adoption of type checking, refer to [this documentation](https://microsoft.github.io/pyright/#/getting-started).

If you want to exclude certain files or directories from the project, you can use the `exclude`, but I don't recommend that approach. Instead, I recommend enabling a few diagnostic checks for your entire project, then gradually add diagnostic rules as you fix type issues in your code. If a particular file is too problematic (e.g. contains significant dynamic code), you can explicitly disable specific diagnostic checks for that file using a [file-level comment](https://microsoft.github.io/pyright/#/comments?id=file-level-type-controls).

This is the approach I've used successfully for gradually adopting typing for several large Python code bases.

**Reply by jrheard-seesaw** [reply]
Interesting! I'll take a step back and start again with that approach. My original plan was to immediately turn the checks up to 11 and then gradually roll out coverage on a file-by-file basis; I hadn't considered rolling out coverage on a check-by-check basis, I'll try that instead. Thanks for the advice!
